deploy-cloud-functions/src/cloudFunctionClient.ts

```diff
@@ -0,0 +1,337 @@
+      getRequest,
+      this.methodOptions,
+    );
+    const currentFunctions: cloudfunctions_v1.Schema$ListFunctionsResponse =
+      currentFunctionsResponse.data;
+    let currentFunctionsNames: string[] = [];
+    if (currentFunctions.functions?.length !== undefined) {
+      currentFunctionsNames = currentFunctions.functions?.map(
+        (currentFunction) => currentFunction.name as string,
+      );
+    }
+    return currentFunctionsNames;
+  }
+
+  /**
+   * Returns a signed URL to upload source code.
+   *
+   * @returns signed url.
+   */
+  async getUploadUrl(): Promise<
+    cloudfunctions_v1.Schema$GenerateUploadUrlResponse
+  > {
+    const authClient = await this.getAuthClient();
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$Generateuploadurl = {
+      parent: this.parent,
+      auth: authClient,
+    };
+    const uploadUrlResponse: GaxiosResponse<cloudfunctions_v1.Schema$GenerateUploadUrlResponse> = await this.gcf.projects.locations.functions.generateUploadUrl(
+      getRequest,
+      this.methodOptions,
+    );
+    return uploadUrlResponse.data;
+  }
+
+  /**
+   * Deploy a CloudFunction.
+   *
+   * @param cf Cloud Function.
+   * @returns CloudFunction deployment response.
+   */
+  async deploy(cf: CloudFunction): Promise<cloudfunctions_v1.Schema$Operation> {
+    const authClient = await this.getAuthClient();
+    const deployedFunctions = await this.getFunctions();
+    const zipPath = await zipDir(cf.sourceDir);
+    const uploadUrl = await this.getUploadUrl();
+    if (!uploadUrl.uploadUrl) {
+      throw new Error('Unable to generate signed Url');
+    }
+    // Upload source code
+    await uploadSource(uploadUrl.uploadUrl, zipPath);
+    // Delete temp zip file after upload
+    await deleteZipFile(zipPath);
+    cf.setSourceUrl(uploadUrl.uploadUrl);
+    // If CF already exists, create a revision else create new deployment
+    if (deployedFunctions.includes(cf.fqn)) {
+      core.info('Creating a function revision');
+      const updateMasks = [
+        'sourceUploadUrl',
+        'name',
+        'environmentVariables',
+        'entryPoint',
+        'runtime',
+        'vpcConnector',
+        'serviceAccountEmail',
+        'timeout',
+        'maxInstances',
+        'eventTrigger.eventType',
+        'eventTrigger.resource',
+        'eventTrigger.service',
+      ];
+      const updateFunctionRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$Patch = {
+        updateMask: updateMasks.join(','),
+        name: cf.fqn,
+        auth: authClient,
+        requestBody: cf.request,
+      };
+      const updateFunctionResponse = await this.gcf.projects.locations.functions.patch(
+        updateFunctionRequest,
+        this.methodOptions,
+      );
+      const awaitUpdate = await this.pollOperation(
```

Comment:
This is needed even though you are awaiting the patch request?

Reply:
I am not sure I follow 😅 
Do you mean why we need `awaitUpdate`?

Reply:
Right, we are already awaiting both the patch and create requests, so I am surprised you have to poll the response for completion.

Reply:
IIUC patch request returns back an [LRO](https://cloud.google.com/functions/docs/reference/rest/v1/operations#Operation) which I need to poll to completion.